[WebGPU] Rename vertex shaders in clamp pipelines to unique names

mwyrzykowski opened the pull request:
#### 8f65768faa781a280fc64bd5a9220bed5f289394
<pre>
[WebGPU] Rename vertex shaders in clamp pipelines to unique names
<a href="https://bugs.webkit.org/show_bug.cgi?id=281903">https://bugs.webkit.org/show_bug.cgi?id=281903</a>
<a href="https://rdar.apple.com/138398308">rdar://138398308</a>

Reviewed by Cameron McCormack and Tadeu Zagallo.

Give better names to the custom vertex functions.

Also use call_once() because otherwise multiple threads could
attempt to create the function simultaneously.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::createSimplePso):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::copyIndexIndirectArgsPipeline):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::runPredispatchIndirectCallValidation):

Canonical link: <a href="https://commits.webkit.org/287698@main">https://commits.webkit.org/287698@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3d4d40fe4151188aee70a04f8151cda3e88ef65

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80454 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59461 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34119 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84975 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31436 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68522 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7763 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62862 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20673 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83523 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52953 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73246 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43165 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50261 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29895 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71402 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27922 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86409 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7680 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5425 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/71236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7855 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69084 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70389 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17557 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14394 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13339 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7642 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7481 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11000 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9286 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -352,15 +352,17 @@ static auto timestampWriteIndex(auto writeIndex)
         ALLOW_DEPRECATED_DECLARATIONS_END
-        id<MTLLibrary> library = [device newLibraryWithSource:@"[[vertex]] float4 vs() { return (float4)0; }" options:options error:&error];
+        id<MTLLibrary> library = [device newLibraryWithSource:@"[[vertex]] float4 vsNop() { return (float4)0; }" options:options error:&error];
         ASSERT_UNUSED(error, !error);
```
Should we log the error here, like for the other predefined shaders we compile?

heycam commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -961,10 +967,10 @@ static void setOwnerWithIdentity(id<MTLResourceSPI> resource, auto webProcessID)
 
-    [[vertex]] void vsUshort(device const IndexDataUshort* indexData [[buffer(0)]],
+    [[vertex]] void vsUshortICB(device const IndexDataUshort* indexData [[buffer(0)]],
         device ICBContainer *icb_container [[buffer(1)]], // <-- must match Device::bufferIndexForICBContainer()
```
Nit, not for this patch: since these are string literals, you could #define the bufferIndexForICBContainer value to re-use in the shader here, instead of relying on the comments.

mwyrzykowski replied:
oh that is a good idea

mwyrzykowski committed:
[WebGPU] Rename vertex shaders in clamp pipelines to unique names
https://bugs.webkit.org/show_bug.cgi?id=281903
rdar://138398308

Reviewed by Cameron McCormack and Tadeu Zagallo.

Give better names to the custom vertex functions.

Also use call_once() because otherwise multiple threads could
attempt to create the function simultaneously.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::createSimplePso):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::copyIndexIndirectArgsPipeline):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::runPredispatchIndirectCallValidation):

Canonical link: https://commits.webkit.org/287698@main
